### PR TITLE
[14/N] Refactor _new_process_group_helper() to remove repeated code

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88330 Allow Process Group to support multiple backends and move PG specfic implementations to backend class
* #87088 Remove ProcessGroupRoundRobin
* **#88351 [14/N] Refactor _new_process_group_helper() to remove repeated code**
* #86409 [13/N] Update gather with CPU/CUDA implementations
* #86408 [12/N] Update scatter with CPU/CUDA implementations

### Changes
- refactor parts of `_new_process_group_helper()` to remove repeated code

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D41188274](https://our.internmc.facebook.com/intern/diff/D41188274)

### CODE REVIEW

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -968,9 +903,32 @@ def _new_process_group_helper(
...
 
                 pg = creator_fn(dist_backend_opts, pg_options)
-            _pg_map[pg] = (backend, store)
-            _pg_names[pg] = group_name
 
+        # Process group wrapper initialization for supported PGs when TORCH_DISTRIBUTED_DEBUG is set
+        if backend in [Backend.GLOO, Backend.NCCL, Backend.UCC]:
+            # In debug mode and if GLOO is available, wrap in a wrapper PG that
+            # enables enhanced collective checking for debugability.
+            if get_debug_level() == DebugLevel.DETAIL:
```
Comment (kwen2501) [inline torch/distributed/distributed_c10d.py:911]: This `if` condition should be put out-most.

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -968,9 +903,32 @@ def _new_process_group_helper(
...
+                if not _GLOO_AVAILABLE:
+                    logger.info(
+                        """TORCH_DISTRIBUTED_DEBUG was set to DETAIL, but
+                                GLOO is not available. Build with Gloo to
+                                create a wrapper process group in debug mode
+                                to aid collective desynchronization debugging."""
+                    )
```
Comment (kwen2501) [inline torch/distributed/distributed_c10d.py:918]: Move the current out-most condition as an `elif` here:
```
elif backend not in [Backend.GLOO, Backend.NCCL, Backend.UCC]:
  logger.info(
    """TORCH_DISTRIBUTED_DEBUG was set to DETAIL, but the current backend does not support this feature."""
  )
else:
```

**Review by kwen2501 — APPROVED** [review]
LGTM. Thanks for refactoring.
Just added minor comments.

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -1047,9 +982,32 @@ def _new_process_group_helper(
                 dist_backend_opts.global_ranks_in_group = global_ranks_in_group
 
                 pg = creator_fn(dist_backend_opts, pg_options)
-            _world.pg_map[pg] = (backend, store)
-            _world.pg_names[pg] = group_name
 
+    # Process group wrapper initialization for supported PGs when TORCH_DISTRIBUTED_DEBUG is set
+    if backend in [Backend.GLOO, Backend.NCCL, Backend.UCC]:
+        # In debug mode and if GLOO is available, wrap in a wrapper PG that
+        # enables enhanced collective checking for debugability.
```
Comment (awgu) [inline torch/distributed/distributed_c10d.py:989]: nit: 
```suggestion
        # enables enhanced collective checking for debuggability.
```
https://wikidiff.com/debugability/debuggability

**Review by awgu — COMMENTED** [review]
Pushing commits as well I see...

File: torch/distributed/distributed_c10d.py (hunk 3 above)
Comment (H-Huang) [reply torch/distributed/distributed_c10d.py:989]: thank you sir

File: torch/distributed/distributed_c10d.py (hunk 2 above)
Comment (H-Huang) [reply torch/distributed/distributed_c10d.py]: Did we want to add a log for this? We currently don't have a message. I am also not sure if there are things other than `ProcessGroupWrapper` that is available when `dist.DebugLevel.DETAIL` is set, maybe the logs are also more verbose so maybe we should just say the current backend does not support `ProcessGroupWrapper`?

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41188274).

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41188274).

**Pull request closed without merging**